### ISSUE TITLE
[ML] Switch unit testing framework from cppunit to boost

### ISSUE THREAD

**Issue by hendrikmuhs** [elastic/ml-cpp#156]
## Motivation

 - moving away from cppunit:
   - quite outdated and on low maintenance (but still used by e.g. libreoffice)
   - requires a lot of boiler plate code, easy to be forgotten
 - boost unit test framework
   - is well maintained
   - already part of the build env (almost, see [*])
   - is simpler to maintain:
       - better macros
       - automatic test case collection
        - much less boilerplate (no header files, no suite classes required)
        - => less error prone
   - is  more feature rich: more and better assertions
  
## Alternatives

Apart from staying with cppunit the most obvious alternative is `gtest`:

  - equally feature rich to boost, at least for the _core_ features we need
  - would be a new dependency

On the long run we might want to make use of mocking libraries which can be combined with both `boost` and `gtest`

## Con's

 [*] After investigation it turned out that we skip building the boost unit test libraries, therefore the build environment isn't ready yet. Rebuilding the images is substantial work.

## Tasks

 - [x] rebuild build environment images
 - [x] migrate tests to boost framework
 - [ ] ensure CI readiness: timings, reporting
 - [x] cleanup/remove cppunit

**Comment by droberts195** [comment]
The reason this is labeled `v7.0.0` is that 6.x is using different compiler versions and we'd have to update the build machine images twice if we were to backport this to 6.x.  It saves some time to only update the build images for the current master branch environment.  However, if the changes to the test suites were made in master now then they would cause considerable backporting pain for the remaining development of 6.x.  Therefore it would be best to do the majority of the changes after feature freeze for 6.6 (ideally while master is still for 7.0 but even if it's after the 7.0 branch has been created the backport of the test framework changes will still be a trivial cherry-pick at that early stage).

**Comment by droberts195** [comment]
We discussed this again in the EAH last week.

We decided that since 6.7 will be maintained for a long time we'd like to backport the changes to the 6.7 branch.  This implies that the best time to make the change would be between 6.6 feature freeze and 6.7 feature freeze.

This also implies that the Boost unit test library needs to be added to the Boost build for the 6.7 toolchain as well as later versions.
